File: test/extensions/dynamic_modules/test_data/go/udp_integration_test/udp_integration_test.go
// UDP listener filter integration test module.
//
// Two filters:
//
//	"test_filter"      — passthrough; returns Continue on every datagram.
//	"stop_iteration"   — drops every datagram by returning StopIteration.
//
// The integration driver sends UDP datagrams through Envoy's udp_proxy and asserts that
// passthrough datagrams reach the upstream while stop_iteration drops them.
package main

import (
	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterUdpListenerFilterConfigFactories(map[string]shared.UdpListenerFilterConfigFactory{
		"test_filter":    &passthroughConfigFactory{},
		"stop_iteration": &stopIterationConfigFactory{},
	})
}

func main() {} //nolint:all

// =============================================================================
// passthrough — Continue
// =============================================================================

type passthroughConfigFactory struct {
	shared.EmptyUdpListenerFilterConfigFactory
}

func (passthroughConfigFactory) Create(_ shared.UdpListenerFilterConfigHandle, _ []byte) (shared.UdpListenerFilterFactory, error) {
	return &passthroughFactory{}, nil
}

type passthroughFactory struct {
	shared.EmptyUdpListenerFilterFactory
}

func (*passthroughFactory) Create(_ shared.UdpListenerFilterHandle) shared.UdpListenerFilter {
	return &passthroughFilter{}
}

type passthroughFilter struct {
	shared.EmptyUdpListenerFilter
}

func (*passthroughFilter) OnData(handle shared.UdpListenerFilterHandle) shared.UdpListenerFilterStatus {
	// Exercise a couple of read-only handle accessors so they're hit at runtime — these
	// methods would crash on bad cgo handling rather than return wrong data.
	_ = handle.GetDatagramSize()
	_, _, _ = handle.GetPeerAddress()
	return shared.UdpListenerFilterStatusContinue
}

// =============================================================================
// stop_iteration — StopIteration
// =============================================================================

type stopIterationConfigFactory struct {
	shared.EmptyUdpListenerFilterConfigFactory
}

func (stopIterationConfigFactory) Create(_ shared.UdpListenerFilterConfigHandle, _ []byte) (shared.UdpListenerFilterFactory, error) {
	return &stopIterationFactory{}, nil
}

type stopIterationFactory struct {
	shared.EmptyUdpListenerFilterFactory
}

func (*stopIterationFactory) Create(_ shared.UdpListenerFilterHandle) shared.UdpListenerFilter {
	return &stopIterationFilter{}
}

type stopIterationFilter struct {
	shared.EmptyUdpListenerFilter
}

func (*stopIterationFilter) OnData(_ shared.UdpListenerFilterHandle) shared.UdpListenerFilterStatus {
	return shared.UdpListenerFilterStatusStopIteration
}

File: test/extensions/dynamic_modules/test_data/rust/udp_integration_test.rs
//! UDP listener filter integration test module. Mirror of
//! test_data/go/udp_integration_test/udp_integration_test.go.
//!
//! Two filters:
//!   "test_filter"    — passthrough; returns Continue on every datagram.
//!   "stop_iteration" — drops every datagram by returning StopIteration.

use envoy_proxy_dynamic_modules_rust_sdk::*;

declare_udp_listener_filter_init_functions!(init, new_filter_config);

fn init() -> bool {
  true
}

fn new_filter_config<EC: EnvoyUdpListenerFilterConfig, ELF: EnvoyUdpListenerFilter>(
  _envoy: &mut EC,
  name: &str,
  _config: &[u8],
) -> Option<Box<dyn UdpListenerFilterConfig<ELF>>> {
  match name {
    "test_filter" => Some(Box::new(PassthroughConfig {})),
    "stop_iteration" => Some(Box::new(StopIterationConfig {})),
    _ => None,
  }
}

// -----------------------------------------------------------------------------
// passthrough
// -----------------------------------------------------------------------------

struct PassthroughConfig {}

impl<ELF: EnvoyUdpListenerFilter> UdpListenerFilterConfig<ELF> for PassthroughConfig {
  fn new_udp_listener_filter(&self, _envoy: &mut ELF) -> Box<dyn UdpListenerFilter<ELF>> {
    Box::new(PassthroughFilter {})
  }
}

struct PassthroughFilter {}

impl<ELF: EnvoyUdpListenerFilter> UdpListenerFilter<ELF> for PassthroughFilter {
  fn on_data(
    &mut self,
    envoy: &mut ELF,
  ) -> abi::envoy_dynamic_module_type_on_udp_listener_filter_status {
    // Exercise a couple of read-only accessors so they're hit at runtime.
    let _ = envoy.get_datagram_data();
    let _ = envoy.get_peer_address();
    abi::envoy_dynamic_module_type_on_udp_listener_filter_status::Continue
  }
}

// -----------------------------------------------------------------------------
// stop_iteration
// -----------------------------------------------------------------------------

struct StopIterationConfig {}

impl<ELF: EnvoyUdpListenerFilter> UdpListenerFilterConfig<ELF> for StopIterationConfig {
  fn new_udp_listener_filter(&self, _envoy: &mut ELF) -> Box<dyn UdpListenerFilter<ELF>> {
    Box::new(StopIterationFilter {})
  }
}

struct StopIterationFilter {}

impl<ELF: EnvoyUdpListenerFilter> UdpListenerFilter<ELF> for StopIterationFilter {
  fn on_data(
    &mut self,
    _envoy: &mut ELF,
  ) -> abi::envoy_dynamic_module_type_on_udp_listener_filter_status {
    abi::envoy_dynamic_module_type_on_udp_listener_filter_status::StopIteration
  }
}
